File: scripts/check_test_coverage_diff.py
#!/usr/bin/env python3
"""Coverage forcing-function — WARN-only (R3 of road-to-test-and-gate-integrity).

A lightweight coverage-governance nudge: warn (never block) when a PR adds a
new gate script without a matching test, so the +0-tests pattern (three
releases shipped behaviour with no coverage) becomes visible at PR time. This
is partly a *social* problem — R3 is a nudge, not a cure.

Honest framing (per the roadmap critique): R3 IS a lightweight coverage-
governance layer, NOT pure hardening. It ships WARN-only, is calibrated from
the REAL Phase 2 backfill experience (not guessed up front), and carries a
sunset clause. The hard-fail flip is a SEPARATE, later decision gated on
warn-phase data.

----------------------------------------------------------------------------
TRIGGER SURFACE — calibrated from Phase 2 evidence + AI council
(claude-sonnet-4-5 + gpt-4o, analysis, 2026-06-02):

  WARN when a NEW `scripts/check_*.py` or `scripts/lint_*.py` gate is added
  with no matching new/changed test in the same diff (`tests/**/test_*.py`
  or `tests/**/*_test.py` whose stem matches the gate's stem).

What is DELIBERATELY excluded (the council killed these to avoid
pragma-spam / signal-to-noise collapse — "over-broad triggers train people
to bypass it"):
  - Edits to EXISTING gate scripts. Phase 2 migrated 4 gates with one-line
    path-constant swaps that legitimately needed NO test; a trigger that
    fired on edits would be pure noise. KNOWN RECALL LIMIT: a behaviour
    change to an existing gate ships untracked in this warn-only phase —
    accepted on purpose; revisit only if the warn-phase data justifies it.
  - New skills with `requires_skills`. Skills are markdown capability docs,
    usually test-less by design → would fire constantly. Dropped for Phase 1
    (council: "no Phase 2 evidence; immediate pragma-spam").
  - Taskfile wiring, docs/`*.md`, `agents/**`, `config/**`, logging/path-
    constant edits.

PRAGMA — `# coverage-diff-ignore: <reason>` (reason mandatory) in the new
gate file suppresses its warning. Because the trigger is NEW files only, the
pragma lives in the added lines (it IS the diff), so it is naturally
commit-scoped — it cannot silently silence FUTURE edits to the file (those
do not trigger). Mirrors the `check-refs` `<!-- ref-ignore -->` allowlist.

SUCCESS METRIC + SUNSET (the fail-mode flip is a separate later decision):
  - Track warn-rate and pragma-rate over the next release cycle (this script
    emits `coverage-diff: warned=N suppressed=M` so the data is in CI logs —
    no new telemetry surface).
  - Consider fail-mode only at >= 3 legitimate catches per cycle AND
    pragma-rate < 10%.
  - If precision is poor after one cycle, REMOVE the check rather than
    escalate it. A noisy nudge people route around is worse than none.
----------------------------------------------------------------------------

Usage:
  python3 scripts/check_test_coverage_diff.py [--base-ref REF] [--files-status "A\tpath" ...]
Exit code: always 0 (warn-only by contract this phase).
"""
from __future__ import annotations

import argparse
import re
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

GATE_RE = re.compile(r"^scripts/(?:check_|lint_)[A-Za-z0-9_]+\.py$")
PRAGMA_RE = re.compile(r"#\s*coverage-diff-ignore:\s*(\S.*?)\s*$")
_PRAGMA_SCAN_LINES = 60


def _is_test_file(path: str) -> bool:
    if not (path.startswith("tests/") and path.endswith(".py")):
        return False
    stem = Path(path).stem
    return stem.startswith("test_") or stem.endswith("_test")


def _test_matches_gate(gate_path: str, test_paths: list[str]) -> bool:
    """A test covers a gate when its stem shares the gate's stem.

    Accepts naming variance: `tests/test_check_foo.py`, `tests/foo_test.py`,
    `tests/sub/test_foo.py` all match gate `scripts/check_foo.py`.
    """
    gate_stem = Path(gate_path).stem  # e.g. check_foo
    short = gate_stem.removeprefix("check_").removeprefix("lint_")  # foo
    for t in test_paths:
        tstem = Path(t).stem.removeprefix("test_").removesuffix("_test")
        if gate_stem in Path(t).stem or short and short in tstem:
            return True
    return False


def evaluate(changed, pragma_reason):
    """Pure core (no git, no I/O): classify the changed-file set.

    `changed` is a list of (status, path) where status is the git
    name-status code ('A' added, 'M' modified, …). `pragma_reason(path)`
    returns the in-file opt-out reason or None. Returns
    `(warnings, suppressed)` where warnings is a list of new gate paths
    with no matching test and no pragma, suppressed is [(path, reason)].
    """
    new_gates = [p for (s, p) in changed if s == "A" and GATE_RE.match(p)]
    test_changes = [p for (_s, p) in changed if _is_test_file(p)]
    warnings: list[str] = []
    suppressed: list[tuple[str, str]] = []
    for gate in new_gates:
        if _test_matches_gate(gate, test_changes):
            continue
        reason = pragma_reason(gate)
        if reason:
            suppressed.append((gate, reason))
        else:
            warnings.append(gate)
    return warnings, suppressed


def _pragma_reason_from_tree(path: str) -> str | None:
    f = REPO_ROOT / path
    try:
        head = f.read_text(encoding="utf-8", errors="ignore").splitlines()[:_PRAGMA_SCAN_LINES]
    except OSError:
        return None
    for line in head:
        m = PRAGMA_RE.search(line)
        if m:
            return m.group(1)
    return None


def _resolve_base_ref(explicit: str | None) -> str:
    if explicit:
        return explicit
    for candidate in ("origin/main", "origin/master", "main", "master"):
        try:
            subprocess.check_output(
                ["git", "rev-parse", "--verify", candidate], stderr=subprocess.DEVNULL,
            )
            return candidate
        except subprocess.CalledProcessError:
            continue
    return "HEAD~1"


def _git_name_status(base_ref: str) -> list[tuple[str, str]]:
    try:
        out = subprocess.check_output(
            ["git", "diff", "--name-status", f"{base_ref}...HEAD"],
            stderr=subprocess.STDOUT, text=True,
        )
    except subprocess.CalledProcessError as exc:
        print(f"⚠️  coverage-diff: git diff failed ({exc.output.strip()}); skipping.", file=sys.stderr)
        return []
    rows: list[tuple[str, str]] = []
    for line in out.splitlines():
        parts = line.split("\t")
        if len(parts) >= 2:
            rows.append((parts[0][:1], parts[-1]))
    return rows


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--base-ref", default=None)
    opts = ap.parse_args(argv)
    changed = _git_name_status(_resolve_base_ref(opts.base_ref))
    warnings, suppressed = evaluate(changed, _pragma_reason_from_tree)
    if warnings:
        print("⚠️  coverage-diff: new gate(s) added with no matching test (warn-only):")
        for g in warnings:
            print(f"    {g} — add tests/test_{Path(g).stem}.py, or a "
                  f"`# coverage-diff-ignore: <reason>` line if no test is warranted.")
    for g, reason in suppressed:
        print(f"    (suppressed: {g} — {reason})")
    print(f"coverage-diff: warned={len(warnings)} suppressed={len(suppressed)}")
    return 0  # warn-only by contract this phase


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/check_test_coverage_diff.ts
#!/usr/bin/env tsx
/**
 * Coverage forcing-function — WARN-only (R3 of road-to-test-and-gate-integrity).
 *
 * TypeScript twin of `src/scripts/check_test_coverage_diff.py` (ADR-094,
 * Phase 4 / Wave 4c). Mirrors the Python CLI contract EXACTLY — `--base-ref`
 * flag, always exit 0 (warn-only by contract), stdout, byte-identical
 * messages, the git name-status read, the pragma scan, and the `evaluate`
 * core logic. No behaviour changes — latent matcher quirks replicated.
 *
 * WARN when a NEW `src/scripts/{check,lint}_*.py` gate is added with no
 * matching new/changed test in the same diff and no in-file pragma.
 */

import { spawnSync } from 'node:child_process';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const _HERE = fileURLToPath(import.meta.url);

// Injectable for tests (monkeypatch parity): REPO_ROOT, _git_name_status,
// _pragma_reason_from_tree.
let REPO_ROOT = path.resolve(path.dirname(_HERE), '..', '..');

const GATE_RE = /^src\/scripts\/(?:check_|lint_)[A-Za-z0-9_]+\.py$/;
const PRAGMA_RE = /#\s*coverage-diff-ignore:\s*(\S.*?)\s*$/;
const _PRAGMA_SCAN_LINES = 60;

type Changed = ReadonlyArray<[string, string]>;
type PragmaReason = (path: string) => string | null;

function _is_test_file(p: string): boolean {
    if (!(p.startsWith('tests/') && p.endsWith('.py'))) {
        return false;
    }
    const stem = _stem(p);
    return stem.startsWith('test_') || stem.endsWith('_test');
}

/** Mirror Python `Path(p).stem` — basename without the final suffix. */
function _stem(p: string): string {
    const base = p.split('/').pop() ?? p;
    const dot = base.lastIndexOf('.');
    return dot > 0 ? base.slice(0, dot) : base;
}

/** Mirror Python `str.removeprefix`. */
function _removeprefix(s: string, prefix: string): string {
    return s.startsWith(prefix) ? s.slice(prefix.length) : s;
}

/** Mirror Python `str.removesuffix`. */
function _removesuffix(s: string, suffix: string): string {
    return s.endsWith(suffix) ? s.slice(0, s.length - suffix.length) : s;
}

function _test_matches_gate(gate_path: string, test_paths: readonly string[]): boolean {
    const gate_stem = _stem(gate_path); // e.g. check_foo
    const short = _removeprefix(_removeprefix(gate_stem, 'check_'), 'lint_'); // foo
    for (const t of test_paths) {
        const tstem = _removesuffix(_removeprefix(_stem(t), 'test_'), '_test');
        // Replicate Python: `gate_stem in Path(t).stem or short and short in tstem`.
        if (_stem(t).includes(gate_stem) || (short !== '' && tstem.includes(short))) {
            return true;
        }
    }
    return false;
}

function evaluate(
    changed: Changed,
    pragma_reason: PragmaReason,
): [string[], Array<[string, string]>] {
    const new_gates = changed.filter(([s, p]) => s === 'A' && GATE_RE.test(p)).map(([, p]) => p);
    const test_changes = changed.filter(([, p]) => _is_test_file(p)).map(([, p]) => p);
    const warnings: string[] = [];
    const suppressed: Array<[string, string]> = [];
    for (const gate of new_gates) {
        if (_test_matches_gate(gate, test_changes)) {
            continue;
        }
        const reason = pragma_reason(gate);
        if (reason) {
            suppressed.push([gate, reason]);
        } else {
            warnings.push(gate);
        }
    }
    return [warnings, suppressed];
}

function _pragma_reason_from_tree(p: string): string | null {
    const f = path.join(REPO_ROOT, p);
    let head: string[];
    try {
        head = fs.readFileSync(f, 'utf-8').split('\n').slice(0, _PRAGMA_SCAN_LINES);
    } catch {
        return null;
    }
    for (const line of head) {
        const m = PRAGMA_RE.exec(line);
        if (m) {
            return m[1]!;
        }
    }
    return null;
}

function _resolve_base_ref(explicit: string | null): string {
    if (explicit) {
        return explicit;
    }
    for (const candidate of ['origin/main', 'origin/master', 'main', 'master']) {
        const res = spawnSync('git', ['rev-parse', '--verify', candidate], {
            stdio: ['ignore', 'ignore', 'ignore'],
        });
        if (res.status === 0) {
            return candidate;
        }
    }
    return 'HEAD~1';
}

function _git_name_status(base_ref: string): Array<[string, string]> {
    const res = spawnSync('git', ['diff', '--name-status', `${base_ref}...HEAD`], {
        encoding: 'utf-8',
    });
    if (res.status !== 0) {
        // Mirror Python: combined stdout+stderr (subprocess.STDOUT) .strip().
        const out = `${res.stdout ?? ''}${res.stderr ?? ''}`.trim();
        process.stderr.write(
            `⚠️  coverage-diff: git diff failed (${out}); skipping.\n`,
        );
        return [];
    }
    const rows: Array<[string, string]> = [];
    for (const line of (res.stdout ?? '').split('\n')) {
        const parts = line.split('\t');
        if (parts.length >= 2) {
            rows.push([parts[0]!.slice(0, 1), parts[parts.length - 1]!]);
        }
    }
    return rows;
}

interface ParsedArgs {
    base_ref: string | null;
}

function _argparse_error(message: string): never {
    process.stderr.write(`check_test_coverage_diff: error: ${message}\n`);
    process.exit(2);
}

function parse_args(argv: readonly string[]): ParsedArgs {
    let base_ref: string | null = null;
    for (let i = 0; i < argv.length; i++) {
        const arg = argv[i]!;
        if (arg === '--base-ref') {
            const v = argv[++i];
            if (v === undefined) {
                _argparse_error('argument --base-ref: expected one argument');
            }
            base_ref = v;
        } else if (arg.startsWith('--base-ref=')) {
            base_ref = arg.slice('--base-ref='.length);
        } else if (arg === '-h' || arg === '--help') {
            process.stdout.write('usage: check_test_coverage_diff [-h] [--base-ref BASE_REF]\n');
            process.exit(0);
        } else {
            _argparse_error(`unrecognized arguments: ${arg}`);
        }
    }
    return { base_ref };
}

function main(argv?: readonly string[]): number {
    const opts = parse_args(argv ?? process.argv.slice(2));
    const changed = _hooks.git_name_status(_resolve_base_ref(opts.base_ref));
    const [warnings, suppressed] = evaluate(changed, _hooks.pragma_reason_from_tree);
    if (warnings.length > 0) {
        process.stdout.write(
            '⚠️  coverage-diff: new gate(s) added with no matching test (warn-only):\n',
        );
        for (const g of warnings) {
            process.stdout.write(
                `    ${g} — add tests/test_${_stem(g)}.py, or a ` +
                    `\`# coverage-diff-ignore: <reason>\` line if no test is warranted.\n`,
            );
        }
    }
    for (const [g, reason] of suppressed) {
        process.stdout.write(`    (suppressed: ${g} — ${reason})\n`);
    }
    process.stdout.write(
        `coverage-diff: warned=${warnings.length} suppressed=${suppressed.length}\n`,
    );
    return 0; // warn-only by contract this phase
}

// Indirection layer so tests can monkeypatch _git_name_status /
// _pragma_reason_from_tree the way pytest does on the Python module.
const _hooks = {
    git_name_status: _git_name_status,
    pragma_reason_from_tree: _pragma_reason_from_tree,
};

function _set_repo_root_for_test(root: string): void {
    REPO_ROOT = root;
}

function _set_hooks_for_test(overrides: Partial<typeof _hooks>): void {
    Object.assign(_hooks, overrides);
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}

export {
    GATE_RE,
    PRAGMA_RE,
    evaluate,
    _pragma_reason_from_tree,
    _git_name_status,
    _test_matches_gate,
    _is_test_file,
    main,
    _set_repo_root_for_test,
    _set_hooks_for_test,
};
